### PR TITLE
feat: sbtc integration

### PR DESCRIPTION
**Pull request by kyranjamie**
> Try out Leather build bfcbe19 — [Extension build](https://github.com/leather-io/extension/actions/runs/12396921673), [Test report](https://leather-io.github.io/playwright-reports/feat/display-sbtc), [Storybook](https://feat/display-sbtc--65982789c7e2278518f189e7.chromatic.com), [Chromatic](https://www.chromatic.com/library?appId=65982789c7e2278518f189e7&branch=feat/display-sbtc)

This PR ensures sBTC displays correctly, with market data.

There are several design variants I'm not entirely sure which we should go for. No icon for now because apparently there's a new one on the way.

I'm a bit uncertain how we should handle sBTC specific behaviour. Either we can treat is separately from other SIP-10s or, manipulate the data/events we pass to the SIP-10 components such that it displays/behaves as we expect. Trying to avoid lots of `if (symbol === 'sBTC')`-like code.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -181,7 +181,7 @@
     "@stacks/profile": "6.15.0",
     "@stacks/rpc-client": "1.0.3",
     "@stacks/storage": "6.15.0",
-    "@stacks/transactions": "6.17.0",
+    "@stacks/transactions": "7.0.2",
```
Comment (fbwoolf) [inline package.json]: I am hitting a lot of errors with `@stacks/transactions` imports with this change, what am I missing?

Comment (fbwoolf) [reply package.json]: All good if I revert it back...

Screenshot 2024-11-30 at 1 23 40 PM

Comment (fbwoolf) [reply package.json]: I see we prob need lib update for the sbtc tx. Still looking thru this so ignore me, haha.

**Comment by markmhendrickson** [comment]
Let's link to an issue on the board?

**Comment by fbwoolf** [comment]
> Let's link to an issue on the board?

I'll look for an issue. I just stepped in to take this over, so I'm not sure what issue it relates to specifically. I think we are just trying to get a POC working for the sBTC deposit tx, and then we can apply that to swaps UI.

**Comment by fbwoolf** [comment]
Just an update here, working branched off on adding all the stuff needed to the swaps UI while we wait on the sbtc pkg update. I think tracking all the changes needed to update the stacks.js pkgs with copying the code over will be quite a chunk of work, so stalling for the lib update from Hiro. I will circle back to tracking the updates if that lib doesn't move fwd in the next few days.

**Comment by fbwoolf** [comment]
sbtc pkg was updated, so I'll update here and hopefully remove copied code 🤞 https://www.npmjs.com/package/sbtc

**Comment by fbwoolf** [comment]
Made the update, but testnet client is still broken. Got to this point thou...

Screenshot 2024-12-03 at 2 03 59 PM

**Comment by fbwoolf** [comment]
Swaps UI draft here: https://github.com/leather-io/extension/pull/5997

**Comment by fbwoolf** [comment]
The sbtc pkg was updated again with new documentation, I think I got the start of the deposit tx working. I believe we just need to add the inputs to it...

Screenshot 2024-12-03 at 7 30 55 PM

**Comment by fbwoolf** [comment]
Ok, got the inputs and signed/finalized tx working... have not tried broadcasting yet.

Screenshot 2024-12-03 at 8 46 09 PM

**Comment by fbwoolf** [comment]
As an update, currently stuck here and waiting for feedback... https://github.com/hirosystems/stacks.js/pull/1554/files#r1870157852

File: src/app/pages/swap/components/swap-assets-pair/swap-asset-item.layout.tsx [hunk 2]
```diff
@@ -1,18 +1,27 @@
...
   symbol: string;
   value: string;
 }
 export function SwapAssetItemLayout({ caption, icon, symbol, value }: SwapAssetItemLayoutProps) {
   return (
     <Flag
-      img={<styled.img src={icon} borderRadius="50%" width="48px" height="48px" alt="Swap asset" />}
+      img={
+        isString(icon) ? (
+          <styled.img src={icon} borderRadius="50%" width="48px" height="48px" alt="Swap asset" />
```
[the remark this answers is not in the document]
Comment (pete-watters) [reply src/app/pages/swap/components/swap-assets-pair/swap-asset-item.layout.tsx]: There's a new 🐰 in town 🫨

**Comment by fbwoolf** [comment]
@kyranjamie we seem to be using both `SBtc...` and `Sbtc` in naming here, what is your pref? When I look up convention, I think `SBtc` is correct?

**Comment by kyranjamie** [comment]
@fbwoolf yeah it's a bit of a contrived name tbh, given it's an abbreviation with a prefix added on. I'd treat it as a single name/word though and go with `Sbtc`, if at beginning of a camel case variable

**Comment by fbwoolf** [comment]
> @fbwoolf yeah it's a bit of a contrived name tbh, given it's an abbreviation with a prefix added on. I'd treat it as a single name/word though and go with `Sbtc`, if at beginning of a camel case variable

Yeah, I wasn't feeling like `SBtc...` looked quite right. I'll coordinate the change. Working on tests atm and cleaning things up.

**Comment by fbwoolf** [comment]
As an update here, for the most part all is done and working but a bit hard to test bc testnet is down. Most of my registered deposits just get stuck pending atm. They are going to ping me once testnet is more reliable. At that point I can test the activity feed a bit better. Swap UI is all good thou and working. Adding tests now and cleaning up. Feel free to leave comments on code if see concerns.

Screenshot 2024-12-10 at 10 37 38 AM

File: tests/specs/swap/swap.spec.ts [hunk 3]
```diff
@@ -34,25 +35,28 @@ test.skip('Swaps', () => {
...
 
-    const request = await requestPromise;
-    const requestBody = request.postDataBuffer();
-    test.expect(requestBody).toBeDefined();
+    const toastMessage = 'Transaction submitted!';
+    const toast = swapPage.page.getByText(toastMessage, { exact: true });
+    await expect(toast).toBeVisible();
+  });
+
+  test.only('that it preselects cross chain swap assets', async ({ swapPage }) => {
```
Comment (kyranjamie) [inline tests/specs/swap/swap.spec.ts]:
```suggestion
  test('that it preselects cross chain swap assets', async ({ swapPage }) => {
```

Comment (fbwoolf) [reply tests/specs/swap/swap.spec.ts]: O gosh, haha, will fix.

**Comment by markmhendrickson** [comment]
This is relatively minor since most users won't navigate directly to this swaps page state or refresh there, but if I reload the page after selecting BTC->sBTC, I'm unable to select "max" to proceed:


https://github.com/user-attachments/assets/be3a00c7-572d-4868-b93b-b806bacb4509

**Comment by markmhendrickson** [comment]
I presume I should be seeing a positive transaction fee value here? It appears as 0 BTC on both mainnet and sBTC testnet, having selected my entire balance the previous screen:

image

image

However, if I enter a manual value in the previous screen (below my total balance), I do see a fee value:

image

**Comment by markmhendrickson** [comment]
Let's add `cursor: pointer` to this banner CTA so the user knows it's clickable:

image

**Comment by markmhendrickson** [comment]
@fabric-8 to confirm, but IIRC we display to place a different banner variant here that prompts the user to bridge BTC to sBTC in the condition in which they have BTC but no sBTC yet:

image

And this banner would link to the swaps module with BTC->sBTC preselected

**Comment by markmhendrickson** [comment]
I'm wondering if we should hide this banner somehow once the user has enrolled (or if they're not interested). We could try to detect on-chain enrollment, though it may be easiest to add a simply dismiss option that saves its hidden state client-side in cache?

image

**Comment by markmhendrickson** [comment]
The BTC transaction to bridge appeared in activity for me, but I'm not seeing a separate bridging activity item on the Stacks side:

image

File: src/app/pages/swap/components/swap-asset-select/components/swap-amount-field.tsx [hunk 4]
```diff
@@ -35,6 +36,13 @@ export function SwapAmountField({ amountAsFiat, isDisabled, name }: SwapAmountFi
   const [field] = useField(name);
   const showError = useShowFieldError(name) && name === 'swapAmountBase' && values.swapAssetQuote;
 
+  useEffect(() => {
+    // Clear quote amount if quote asset is reset
+    if (isUndefined(values.swapAssetQuote)) {
+      void setFieldValue('swapAmountQuote', '');
+    }
+  }, [setFieldValue, values]);
```
Comment (kyranjamie) [inline src/app/pages/swap/components/swap-asset-select/components/swap-amount-field.tsx:44]: I wonder if there are other ways we can do this. Seems kinda side effecty to have this component setting/updating state like this

File: src/app/pages/swap/components/swap-assets-pair/swap-assets-pair.tsx [hunk 5]
```diff
@@ -21,14 +21,22 @@ export function SwapAssetsPair() {
     return null;
   }
 
+  const formattedSwapAmountBase = formatMoneyWithoutSymbol(
+    createMoneyFromDecimal(
+      Number(swapAmountBase),
+      swapAssetBase.name,
+      swapAssetBase.balance.decimals
+    )
+  );
```
Comment (kyranjamie) [inline src/app/pages/swap/components/swap-assets-pair/swap-assets-pair.tsx]:
```suggestion
  const swapAmountBase = createMoneyFromDecimal(
    Number(swapAmountBase),
    swapAssetBase.name,
    swapAssetBase.balance.decimals
  );
```

File: src/app/pages/swap/components/swap-assets-pair/swap-assets-pair.tsx [hunk 6]
```diff
@@ -21,14 +21,22 @@ export function SwapAssetsPair() {
...
+
   return (
     <SwapAssetsPairLayout
       swapAssetBase={
         <SwapAssetItemLayout
           caption="You will swap"
           icon={swapAssetBase.icon}
           symbol={swapAssetBase.name}
-          value={swapAmountBase}
+          value={formattedSwapAmountBase}
```
Comment (kyranjamie) [inline src/app/pages/swap/components/swap-assets-pair/swap-assets-pair.tsx]:
```suggestion
          value={formatMoneyWithoutSymbol(swapAmountBase)}
```

small suggestion, but good idea to format at last minute

File: src/app/pages/swap/hooks/use-bitcoin-bridge-assets.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import { useCurrentAccountNativeSegwitIndexZeroSigner } from '@app/store/accounts/blockchain/bitcoin/native-segwit-account.hooks';
+import { useCurrentStacksAccountAddress } from '@app/store/accounts/blockchain/stacks/stacks-account.hooks';
+
+export function useBtcSwapAsset() {
+  const nativeSegwitSigner = useCurrentAccountNativeSegwitIndexZeroSigner();
+  const currentBitcoinAddress = nativeSegwitSigner.address;
+  const { balance } = useBtcCryptoAssetBalanceNativeSegwit(currentBitcoinAddress);
+  const bitcoinMarketData = useCryptoCurrencyMarketDataMeanAverage('BTC');
+
+  return useCallback((): SwapAsset => {
```
Comment (kyranjamie) [inline src/app/pages/swap/hooks/use-bitcoin-bridge-assets.tsx]: Could be a useMemo?

File: src/app/pages/swap/hooks/use-sbtc-deposit-transaction.tsx [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+import { useBitcoinScureLibNetworkConfig } from '@app/store/accounts/blockchain/bitcoin/bitcoin-keychain';
+import { useCurrentAccountNativeSegwitIndexZeroSigner } from '@app/store/accounts/blockchain/bitcoin/native-segwit-account.hooks';
+import { useCurrentStacksAccount } from '@app/store/accounts/blockchain/stacks/stacks-account.hooks';
+import { useCurrentNetwork } from '@app/store/networks/networks.selectors';
+
+import type { SwapSubmissionData } from '../swap.context';
+
+// Suggested to use as defaults
+const maxSignerFee = 80_000;
+const reclaimLockTime = 6_000;
```
Comment (kyranjamie) [inline src/app/pages/swap/hooks/use-sbtc-deposit-transaction.tsx]: as discussed on call this should be 144?

**Comment by markmhendrickson** [comment]
Let's add "BTC" as a suffix here for max signer fee?

image

**Review by fbwoolf — APPROVED** [review]

**Pull request merged**